numba/tests/test_lists.py

```diff
@@ -865,5 +874,530 @@ def test_reflect_clean(self):
+        from copy import deepcopy
+        expect_args = deepcopy(args)
+        expect = pyfunc(*expect_args)
+
+        njit_args = deepcopy(args)
+        cfunc = jit(nopython=True)(pyfunc)
+        got = cfunc(*njit_args)
+
+        self.assert_list_element_precise_equal(
+            expect=expect, got=got
+            )
+        # Check reflection
+        self.assert_list_element_precise_equal(
+            expect=expect_args, got=njit_args
+            )
+
+    def test_returning_list_of_list(self):
+        def pyfunc():
+            a = [[np.arange(i)] for i in range(4)]
+            return a
+
+        self.compile_and_test(pyfunc)
+
+    @expect_reflection_failure
+    def test_heterogeneous_list_error(self):
+        def pyfunc(x):
+            return x[1]
+
+        cfunc = jit(nopython=True)(pyfunc)
+        l2 = [[np.zeros(i) for i in range(5)],
+              [np.ones(i)+1j for i in range(5)]]
+        l3 = [[np.zeros(i) for i in range(5)], [(1,)]]
+        l4 = [[1], [{1}]]
+        l5 = [[1], [{'a': 1}]]
+
+        # TODO: this triggers a reflection error.
+        # Remove this line when nested reflection is supported
+        cfunc(l2)
+
+        # error_cases
+        with self.assertRaises(TypeError) as raises:
+            cfunc(l2)
+
+        self.assertIn(
+            ("reflected list(array(float64, 1d, C)) != "
+             "reflected list(array(complex128, 1d, C))"),
+            str(raises.exception)
+            )
+
+        with self.assertRaises(TypeError) as raises:
+            cfunc(l3)
+
+        self.assertIn(
+            ("reflected list(array(float64, 1d, C)) != "
+             "reflected list((int64 x 1))"),
+            str(raises.exception)
+            )
+
+        with self.assertRaises(TypeError) as raises:
+            cfunc(l4)
+        self.assertIn(
+            "reflected list(int64) != reflected list(reflected set(int64))",
+            str(raises.exception)
+            )
+
+        with self.assertRaises(ValueError) as raises:
+            cfunc(l5)
+        self.assertIn(
+            "Cannot type list element of <class 'dict'>",
+            str(raises.exception)
+            )
+
+    @expect_reflection_failure
+    def test_list_of_list_reflected(self):
+        def pyfunc(l1, l2):
+            l1.append(l2)
+            l1[-1].append(123)
+
+        cfunc = jit(nopython=True)(pyfunc)
+        l1 = [[0, 1], [2, 3]]
+        l2 = [4, 5]
+        expect = list(l1), list(l2)
+        got = list(l1), list(l2)
+        pyfunc(*expect)
+        cfunc(*got)
+        self.assertEqual(expect, got)
+
+    @expect_reflection_failure
+    def test_heterogenous_list(self):
+        def pyfunc(x):
+            return x[1]
+
+        l1 = [[np.zeros(i) for i in range(5)], [np.ones(i) for i in range(5)]]
+
+        cfunc = jit(nopython=True)(pyfunc)
+        l1_got = cfunc(l1)
+        self.assertPreciseEqual(pyfunc(l1), l1_got)
+
+    @expect_reflection_failure
+    def test_c01(self):
```

Comment:
Thanks for including all these. It adds confidence that a load of common and some less common cases are handled in an expected manner.